modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -231,6 +271,58 @@ public void start() {
         metricManager.enable(PLANNING_EXECUTOR_SOURCE_NAME);
 
         IgnitePlanner.warmup();
+
+        planUpdater.scheduleAtFixedRate(() -> {
+            CompletableFuture<PlanInfo> planFut = rePlanningFut.get();
+            if (planFut != null && !planFut.isDone()) {
+                // some work still in progress
+                return;
+            } else {
+                rePlanningFut.set(null);
+            }
+
+            for (Entry<CacheKey, Object> ent : reScheduledPlans.entrySet()) {
+                CacheKey key = ent.getKey();
+                CompletableFuture<PlanInfo> fut = cache.get(key);
+
+                // can be evicted
+                if (fut != null) {
+                    assert isCompletedSuccessfully(fut);
```

Comment:
Can the future be cancelled due to timeout?

Reply:
no, reScheduledPlans are filled only with successfully completed tasks.